### DISCUSSION TITLE
How to support custom parser in VS Code?

### DISCUSSION THREAD

**Discussion by gentlee** [babel/babel#13599] [Q&A]
Question in SO: https://stackoverflow.com/questions/68518043/support-custom-babel-parser-in-vs-code

I'm making a custom babel parser for my babel plugin where I have a new symbol, it works awesome but VS Code doesn't know about it and keeps showing errors for that symbol.

Is there a way to set custom parser for VS Code or I'm missing something in configuration?

**babel.config.json**
```
{
  "presets": ["@babel/preset-env"],
  "plugins": ["custom-plugin"]
}
```

**custom-plugin**
```
const customParser = require('custom-parser')

const customPlugin = ({ types }) => {
return {
  parserOverride(code, opts) {
    return customParser.parse(code, opts)
  },
  ...
}
```

**Comment by JLHwung** [comment]
> VS Code doesn't know about it and keeps showing errors for that symbol.

Where is the error from? Is that from the builtin `language-javascript` plugin? If so you will have to fork [`language-javascript`](https://github.com/microsoft/vscode/tree/main/extensions/javascript), I don't think it uses Babel.

**Reply by gentlee** [reply]
Yes, looks like.
So there is no good way to write custom plugin that uses custom parser that could be used normally? No sense in writing custom parser?

**Reply by fedeci** [reply]
It makes sense to write a custom parser, but to get your syntax correctly highlighted you should fork (or create your own) the vscode extension responsible of syntax highlighting for `js` files.

**Reply by gentlee** [reply]
1) Who will use plugin that requires custom extension for each IDE and additional configuration? And these extensions also should be constantly updated.

2) How I see it:
 - it should be possible to write plugins for babel parser, they will be automatically supported by IDEs with time
 - It should be possible to set custom parser for babel in babel config, which also will be supported by IDEs with time

**Reply by fedeci** [reply]
It is impossible to make `@babel/parser` directly pluggable because of the parsing conflicts that may occur. Right now the only way to parse a non spec compliant version of javascript is forking the parser. About the IDEs you should ask to them to start using `estree` parsers because at the moment it seems they are using regexes to handle highlighting.

**Reply by gentlee** [reply]
But using `estree` parsers won't solve the issue - there is still no way to at least set default parser for the project in babel.config, isn't it?

**Reply by fedeci** [reply]
[This](https://babeljs.io/docs/en/babel-parser#will-the-babel-parser-support-a-plugin-system) is the current way to use your custom parser from the configs.
